### ISSUE TITLE
Add Callbacks to Mini-Trainer

### ISSUE THREAD

**Issue by RobotSail** [Red-Hat-AI-Innovation-Team/mini_trainer#76]
## Overview

**Goal:** Add Callbacks to Mini-Trainer

**Why:**

We want to facilitate use-cases such as:
* initiating eval jobs within a GPU cluster on checkpoint save
* ending the training loop early based on some condition
* custom metric tracking for OSFT/continual learning workflows
* integrating with Training-Hub's unified callback system

In order to implement this, we would add features in the following order:
1. [ ] implement simple, read-only callbacks
2. [ ] implement callbacks that can control flow of training loop (e.g. stopping the loop early, initiating a checkpoint, etc.)

**Related Issues:**
- [instructlab/training#694](https://github.com/instructlab/training/issues/694) — equivalent implementation for instructlab-training

## Expected usage

Users calling the `run_training` API would directly provide their callbacks to the `TrainingArgs` object through a flat API.

As an example:

```py
from mini_trainer import run_training, TrainingArgs, TorchrunArgs

run_training(
    torch_args=TorchrunArgs(nproc_per_node=8),
    train_args=TrainingArgs(
        model_name_or_path="meta-llama/Llama-3.1-8B-Instruct",
        data_path="./data.jsonl",
        output_dir="./checkpoints",
        on_save=run_eval,
        on_evaluate=end_if_diverged,
        # ... etc
    ),
)
```

## Implementation Details

To implement callbacks in this training library, we have to worry about two major pieces:
1. Taking callbacks that were provided to the `run_training` API and passing them to the `torchrun` subprocesses which run the actual training loop
2. Receiving the callbacks inside of the training loop (child process of `torchrun`) and executing them at the appropriate points without interfering with the training process

We will eventually also want to implement callbacks which can modify what actions are performed by the training loop itself, however this will require some additional effort to ensure proper synchronization of the global process group. These callbacks will not be covered in this issue though.

### The callback API

We can define a general callback as a generic function which accepts some general context as well as a list of args + kwargs:

```python
def my_callback(context, **kwargs) -> None:
    # do stuff
    return
```

Callbacks may also have arguments for certain events, e.g. `on_save` may have the following interface:

```py
def on_save(context, checkpoint_path: str, **kwargs) -> None:
    # do stuff
    return
```

While the `on_evaluate` callback would look like this:

```py
def on_evaluate(context, validation_loss: float, **kwargs) -> None:
    # do stuff
    return
```

**We expect** these callbacks to be entirely self-contained, so that any and all packages are imported **inside** of the callback function and must be present within whatever venv the training loop runs in.

Additionally, since we expect the initial callbacks to be **read-only** and inconsequential to the training loop, the training loop will wrap them as asynchronous tasks which run on the global event loop. For this reason, callbacks should be careful with their usage of async functionality.

> [!WARNING]
> To prevent any unexpected behavior or complex testing scenarios, the callbacks will not be allowed to have their exceptions propagate into the training loop, therefore any exceptions **must** be handled by the callback or they will fail silently.

### Adding callbacks to `TrainingArgs`

To maintain a simple interface, the `TrainingArgs` dataclass (in `src/mini_trainer/training_types.py`) would expose registration for each callback with a flat interface like this:

```py
from mini_trainer import TrainingArgs

args = TrainingArgs(
    model_name_or_path="...",
    data_path="...",
    output_dir="...",
    on_save=evaluate_gsm8k,
    on_evaluate=save_if_improved,
    ...
)
```

The API for these event hooks should also allow providing multiple callbacks:

```py
@dataclass
class TrainingArgs:
    # ...
    on_save: Callable | list[Callable] | None = None
    on_evaluate: Callable | list[Callable] | None = None
    # ... etc
```

### Passing callbacks to `torchrun` procs

In order to execute callbacks during training, they must first be passed between the main process where `run_training` is called and the child processes that are spawned to run the training loop in parallel. When `run_training` is called (in `src/mini_trainer/api_train.py`), it builds a CLI command string from the `TrainingArgs` dataclass and spawns it as a subprocess via `torchrun`.

Therefore we can provide these callbacks to our training script by first serializing them with base64, passing them as string arguments to the CLI, then unserializing and `eval`-ing each callback to register it in the child's memory space. As long as the callback is self-contained, it should execute on the child process just as it would in the main process.

### Registering and executing callbacks

Once they're unserialized, each callback will be registered **on the global main/rank-0 process only** by a callback manager.
This is necessary to avoid race conditions or a flood of evaluation requests from each process after every checkpoint save. We can add this functionality later, but we should avoid it to start so the initial implementation can be straightforward.

In order to manage callbacks, we should have a class like `CallbacksManager` which registers the callbacks and is responsible for providing them with the relevant context when they're called.

The callback manager would have an `on_<event>` method for each of the following events:

- [ ] `on_train_begin`: Called after initialization (after `Checkpointer`/`AsyncStructuredLogger` init), before the main `while` loop (~line 732 in `train.py`)
- [ ] `on_epoch_begin`: Called after `data_loader.sampler.set_epoch(epoch)`, before iterating batches (~line 746)
- [ ] `on_step_begin`: Called after `reset_batch()`, before the gradient accumulation loop (~line 751)
- [ ] `on_before_forward`: Called before `model(**model_inputs)` (~line 768)
- [ ] `on_after_backward`: Called after `loss.backward()` (~line 780)
- [ ] `on_before_optimizer_step`: Called before `take_gradient_step()` (~line 803)
- [ ] `on_after_optimizer_step`: Called after `take_gradient_step()` returns `grad_norm` (~line 804)
- [ ] `on_log`: Called after building `batch_metrics` dict, before `metric_logger.log_sync()` (~line 841)
- [ ] `on_evaluate`: Called after `compute_validation_loss()` (~line 838)
- [ ] `on_save`: Called after `save_model()` (~line 852)
- [ ] `on_step_end`: Called after checkpointing, before stop-condition check (~line 869)
- [ ] `on_epoch_end`: Called after `epoch += 1` and epoch checkpoint (~line 893)
- [ ] `on_train_end`: Called after the final checkpoint, before `train()` returns (~line 903)

Internally, the callbacks should be treated as asynchronous coroutines. When the manager's event hook is triggered, it would create a task for each of the registered callbacks and schedule them for execution on the global event loop. There should be no expectation of reading the return values of the callbacks or that they pass values from one to another.

### Context available to callbacks

The `train()` function has several key variables that should be exposed to callbacks via a context object:

| Variable | Description |
|---|---|
| `step` | Current training step |
| `epoch` | Current epoch |
| `total_samples` | Cumulative samples processed |
| `total_tokens` | Cumulative tokens processed |
| `batch_metrics` | Dict with loss, lr, grad_norm, tokens_per_second, etc. |
| `last_validation_loss` | Most recent validation loss (if validation is enabled) |
| `training_mode` | Current `TrainingMode` (EPOCH, STEP, TOKEN, INFINITE) |
| `output_dir` | Checkpoint output directory |

### Architectural notes

1. **The `train()` function is monolithic** — all lifecycle events are inline in a single function (no class-based Trainer). The `CallbacksManager` would be passed as a parameter to `train()` and its hooks would be called at the appropriate points in the loop.

2. **The `run_training()` API launches a subprocess** (`api_train.py`), so callbacks defined in the calling process cannot be directly passed. Callbacks must be serialized (e.g., base64-encoded source via `inspect.getsource` or `dill`/`cloudpickle`) to cross the subprocess boundary.

3. **The loop is distributed** (FSDP2). Callbacks should fire on rank-0 only by default to avoid race conditions. The context object should expose `is_main_process` for callbacks that need to check.

4. **The `take_gradient_step()` helper** (line 84) bundles clip + step + scheduler step + validation + zero_grad. Inserting hooks around optimizer step means either splitting this function or wrapping it.

5. **The existing `AsyncStructuredLogger`** already provides async JSONL + wandb + mlflow logging. The callback system should coexist with it, not replace it.
